apps/sms/services/js/moz_mobile_message/moz_mobile_message_shim.js

```diff
@@ -0,0 +1,210 @@
+
+/* exported MozMobileMessageShim */
+
+'use strict';
+(function(exports) {
+
+/**
+ * Name of the service for mozMobileMessage API shim
+ * @type {string}
+ */
+
+const SERVICE_NAME = 'mozMobileMessageShim';
+
+/**
+ * Array of the event names that is corresponding to mozMobileMessage events.
+ * @type {Array.<string>}
+ */
+
+const EVENTS = ['received', 'sending', 'sent', 'failed', 'deleted',
+                'readsuccess', 'deliverysuccess'];
+
+/**
+ * Array of method names that need to be exposed for API shim.
+ * @type {Array.<string>}
+ */
+
+const METHODS = ['getMessage', 'retrieveMMS', 'send', 'sendMMS',
+                 'delete', 'markMessageRead', 'getSegmentInfoForText'];
+
+/**
+ * Array of stream names that need to return data (messages/threads) in chunk.
+ * @type {Array.<string>}
+ */
+
+const STREAMS = ['getThreads', 'getMessages'];
+
+const mozMobileMessage = navigator.mozMobileMessage;
+
+var service = null;
+
+var MozMobileMessageShim = {
+  init: function ms_init() {
+    service = bridge.service(SERVICE_NAME);
+
+    EVENTS.forEach((event) => {
+      mozMobileMessage.addEventListener(
+        event,
+        this[Utils.camelCase(`on-${event}`)]
+      );
+    });
+
+    METHODS.forEach((shimMethod) => {
+      service.method(shimMethod, this[shimMethod]);
+    });
+
+    STREAMS.forEach((shimStream) => {
+      service.stream(shimStream, this[shimStream]);
+    });
+  },
+
+  /* Events */
+
+  onSending: function ms_onSending(e) {
+    service.broadcast('message-sending', { message: e.message });
+  },
+
+  onFailed: function ms_onFailed(e) {
+    service.broadcast('message-failed-to-send', { message: e.message });
+  },
+
+  onDeliverysuccess: function ms_onDeliverysuccess(e) {
+    service.broadcast('message-delivered', { message: e.message });
+  },
+
+  onReadsuccess: function ms_onReadsuccess(e) {
+    service.broadcast('message-read', { message: e.message });
+  },
+
+  onSent: function ms_onSent(e) {
+    service.broadcast('message-sent', { message: e.message });
+  },
+
+  onReceived: function ms_onReceived(e) {
+    service.broadcast('message-received', { message: e.message });
+  },
+
+  onDeleted: function ms_onDeleted(e) {
+    if (e.deletedThreadIds && e.deletedThreadIds.length) {
+      service.broadcast('threads-deleted', {
+        ids: e.deletedThreadIds
+      });
+    }
+  },
+
+  /* Method */  
+
+  getMessage: function ms_getMessag(id) {
+    return mozMobileMessage.getMessage(id);
+  },
+
+  retrieveMMS: function ms_retrieveMMS(id) {
+    return mozMobileMessage.retrieveMMS(id);
+  },
+
+  send: function ms_send(recipient, content, sendOpts) {
+    return mozMobileMessage.send(recipient, content, sendOpts);
+  },
+
+  sendMMS: function ms_sendMMS(dataOpts, sendOpts) {
+    return mozMobileMessage.sendMMS(dataOpts, sendOpts);
+  },
+
+  delete: function ms_delete(id) {
+    return mozMobileMessage.delete(id);
+  },
+
+  markMessageRead: function ms_markMessageRead(id, isRead, returnReq) {
```

Comment:
Can we rename ```returnReq``` to ```sendReadReport``` like it's named in webidl [1]?

[1] http://mxr.mozilla.org/mozilla-central/source/dom/webidl/MozMobileMessageManager.webidl#183